src/platform/plugins/shared/discover/public/application/main/state_management/redux/actions/save_discover_session.ts

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { v4 as uuidv4 } from 'uuid';
+import type { DiscoverSessionTab } from '@kbn/saved-search-plugin/common';
+import type {
+  SaveDiscoverSessionOptions,
+  SaveDiscoverSessionParams,
+} from '@kbn/saved-search-plugin/public';
+import { updateFilterReferences } from '@kbn/es-query';
+import type { DataViewSpec } from '@kbn/data-views-plugin/public';
+import { i18n } from '@kbn/i18n';
+import { cloneDeep, isObject } from 'lodash';
+import { selectAllTabs, selectRecentlyClosedTabs, selectTab } from '../selectors';
+import { createInternalStateAsyncThunk } from '../utils';
+import { selectTabRuntimeState } from '../runtime_state';
+import { internalStateSlice } from '../internal_state';
+import {
+  fromSavedObjectTabToSavedSearch,
+  fromSavedObjectTabToTabState,
+  fromSavedSearchToSavedObjectTab,
+  fromTabStateToSavedObjectTab,
+} from '../tab_mapping_utils';
+import { appendAdHocDataViews, replaceAdHocDataViewWithId, setDataView } from './data_views';
+import { setTabs } from './tabs';
+
+type AdHocDataViewAction = 'copy' | 'replace';
+
+export const saveDiscoverSession = createInternalStateAsyncThunk(
+  'internalState/saveDiscoverSession',
+  async (
+    {
+      newTitle,
+      newCopyOnSave,
+      newTimeRestore,
+      newDescription,
+      newTags,
+      isTitleDuplicateConfirmed,
+      onTitleDuplicate,
+    }: {
+      newTitle: string;
+      newTimeRestore: boolean;
+      newCopyOnSave: boolean;
+      newDescription: string;
+      newTags: string[];
+      isTitleDuplicateConfirmed: boolean;
+      onTitleDuplicate: () => void;
+    },
+    { dispatch, getState, extra: { services, runtimeStateManager } }
+  ) => {
+    const state = getState();
+    const currentTabs = selectAllTabs(state);
+    const adHocDataViews = new Map<
+      string,
+      {
+        dataViewSpec: DataViewSpec;
+        action: AdHocDataViewAction;
+        tabs: DiscoverSessionTab[];
+      }
+    >();
```

Comment:
The ad hoc data views logic here is pretty gnarly, but I'm not sure of a good alternative. We basically need a way to handle cloning default profile data views and replacing custom ones when `copyOnSave` is `true`, but this becomes pretty complicated when dealing with multiple tabs which can each contain their own instance of a data view. The approach I went with was to store an identical copy of each cloned/replaced data view per tab to avoid distinct versions in each tab, which are then deduplicated at runtime when the session is loaded. There may some structural changes to how we store ad hoc data views in Discover sessions in the future that could simplify this.